doc/design/error_handling_requirements.md

```diff
@@ -0,0 +1,228 @@
+    - error codes shall be optional (if none is used, it is mapped to some default error code)
+1. The EH shall support multiple error levels.
+    - the individual error levels are defined by the platform
+    - there has to be at least a FATAL error level (most severe error level)
+    - the error levels shall be ordered from least severe to most severe
+    - if the code requires an ERROR level that is not supported by the platform it shall not compile
+1. The EH shall support a non-recoverable error handling strategy in case of FATAL errors.
+    - non-recoverable errors shall call terminate after optional handling and logging code has been executed
+1. The EH shall support a recoverable error handling strategies for non-FATAL errors.
+    - recoverable error handling shall support execution of sufficiently general user actons (callbacks)
+1. The detection mechanism of the EH shall allow logging the error.
+    - user defined-logging shall be supported
+    - there shall be a default implementaton for logging
+1. The EH shall allow to trace the error to its source location (file, function, line).
+1. The EH shall be disableable for non-FATAL errors.
+    - disabling shall be possible up to some level at compile time; this level shall be non-FATAL
+    - can be used if there is sufficient trust in the system to disable e.g. warning checks
+    - by default all levels defined by the platform are enabled
+    - disabling shall cause the affected error logic in the code to cause no runtime overhead
+
+### Technical Constraints
+
+1. The EH shall be uniquely defined for each platform.
+    - only one error handler can be active for a platform at any time
+    - the EH shall not be changed while the system runs
+    - the EH must be configurable at runtime in the init phase
+      (for tests, but it might be sufficient to change it at compile time if only required for tests)
+    - for tests it must be possible to override the FATAL error reaction causing termination
+    - If there is no error handler set the system shall not run (check at the very beginning or enforc)
+2. The EH shall not use any exceptions.
+    - this excludes user callbacks (impossible to enforce)
+3. The EH shall not use dynamic memory.
+    - this excludes user callbacks (impossible to enforce)
+4. The EH shall be non-blocking.
+    - recoverable errors shall eventually pass back control to the invoking function
+    - cannot be enforced with user callbacks (the user is responsible to use non-blocking calls only)
+5. The EH shall be highly-visible in the code to be able to distinguish it from ordinary function calls.
+    - allows simple search for error handling locations in the code
+    - if implemented via macros it can easily be disabled
+6. The EH may depend on logging to some extent.
+    - only if it is supposed to use the logging capabilities in some platform implementation
+    - the logger cannot use the (complete) EH due to this
```

Comment:
Thread-safety is missing. But apart from the potential logger use this is no problem as the calls need no synchronization in the general case. Synchronization may arise in user-defined error-handling if a error data-base is to be accessed under mutex or similar. But this would happen in the configurable use code (some handler) or some default implementation and hence be the responsibility of this extension code.